### PR TITLE
[WIP] mgr/rook: persist applied drive groups and fix drive_group_loop

### PR DESCRIPTION
**Pull request by josephsawaya**
This PR changes OSD creation to store the applied drive_groups in the
mon's KV store and keeps track of what drive_groups have been stored with
a list that's accessible with the key "drive_group_list". This commit also
changes the drive_group_loop to use those stored drive groups to re-apply
the drive groups periodically.

Signed-off-by: Joseph Sawaya <jsawaya@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/rook/rook_cluster.py [hunk 1]
```diff
@@ -1034,18 +1073,23 @@ def add_osds(self, drive_group, matching_hosts):
             self.creator = LSOCreator(inventory, self.coreV1_api, self.storage_class)
```
Comment (liewegas) [inline src/pybind/mgr/rook/rook_cluster.py:1073]: this is old code, but no reason for `creator` to be part of `self`.. just make it a local variable

**Comment by jmolmo** [comment]
What happens if a drive group definition is changed? 

Example:
In a cluster with 6 devices of 150G and 6 devices of 100G
```
osds.yaml
------------------------------------
service_type: osd
service_id:  myosds
placement:
  host_pattern: '*'
data_devices:
  size: '100G:200G'
------------------------------------
# ceph orch apply -i osds.yml  <--- result in 6 OSD using the 150 G devices
```
Now the user decides that he really wants to use smaller disks, and modify and reapply the drive group
```
------------------------------------
service_type: osd
service_id:  myosds
placement:
  host_pattern: '*'
data_devices:
  size: ':100G'
------------------------------------
# ceph orch apply -i osds.yml  <--- result in 6 additional OSDs using the 150Gb devices
```.

But .. probably the user is going to expect that the previos 6 OSDs of 150 GB will be deleted.... I think that is not going to happen.

**Comment by josephsawaya** [comment]
> What happens if a drive group definition is changed?
> 
> Example:
> In a cluster with 6 devices of 150G and 6 devices of 100G
> 
> ```
> osds.yaml
> ------------------------------------
> service_type: osd
> service_id:  myosds
> placement:
>   host_pattern: '*'
> data_devices:
>   size: '100G:200G'
> ------------------------------------
> # ceph orch apply -i osds.yml  <--- result in 6 OSD using the 150 G devices
> ```
> 
> Now the user decides that he really wants to use smaller disks, and modify and reapply the drive group
> 
> ```
> ------------------------------------
> service_type: osd
> service_id:  myosds
> placement:
>   host_pattern: '*'
> data_devices:
>   size: ':100G'
> ------------------------------------
> # ceph orch apply -i osds.yml  <--- result in 6 additional OSDs using the 150Gb devices
> ```.
> 
> But .. probably the user is going to expect that the previos 6 OSDs of 150 GB will be deleted.... I think that is not going to happen.
> ```

Applying drive groups should only create new OSDs and not remove any of the existing ones so in this example the 6 OSDs that were created on 150G devices would stay up and the orchestrator would just create new OSDs on any device that's 100G

**Comment by liewegas** [comment]
replaced by https://github.com/ceph/ceph/pull/43775

**Pull request closed without merging**